server/registry/actions_artifacts.go

```diff
@@ -51,88 +51,83 @@ func parseArtifactParent(name string) (artifactParent, error) {
 
 // CreateArtifact handles the corresponding API request.
 func (s *RegistryServer) CreateArtifact(ctx context.Context, req *rpc.CreateArtifactRequest) (*rpc.Artifact, error) {
-	db, err := s.getStorageClient(ctx)
+	// Parent name must be valid.
+	parent, err := parseArtifactParent(req.GetParent())
 	if err != nil {
-		return nil, status.Error(codes.Unavailable, err.Error())
+		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-
+	// Artifact body must be nonempty.
 	if req.GetArtifact() == nil {
 		return nil, status.Errorf(codes.InvalidArgument, "invalid artifact %+v: body must be provided", req.GetArtifact())
 	}
-
-	parent, err := parseArtifactParent(req.GetParent())
-	if err != nil {
-		return nil, status.Error(codes.InvalidArgument, err.Error())
-	}
-
+	// Artifact name must be valid.
 	name := parent.Artifact(req.GetArtifactId())
-
 	if err := name.Validate(); err != nil {
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
-
-	// Creation should only succeed when the parent exists.
-	switch parent := parent.(type) {
-	case names.Project:
-		if _, err := db.GetProject(ctx, parent); err != nil {
-			return nil, err
-		}
-	case names.Api:
-		if _, err := db.GetApi(ctx, parent); err != nil {
-			return nil, err
+	var response *rpc.Artifact
+	if err := s.runWithTransaction(ctx, func(ctx context.Context, db *storage.Client) error {
+		// Creation should only succeed when the parent exists.
+		switch parent := parent.(type) {
```

Comment:
nit: Probably subjective, but might this be easier to read like this?

		var err error
		switch parent := parent.(type) {
		case names.Project:
			_, err = db.GetProject(ctx, parent)
		case names.Api:
			_, err = db.GetApi(ctx, parent)
		case names.Version:
			_, err = db.GetVersion(ctx, parent)
		case names.Spec:
			_, err = db.GetSpec(ctx, parent)
		case names.Deployment:
			_, err = db.GetDeployment(ctx, parent)
		}
		if err != nil {
			return err
		}

Reply:
Agreed, thanks.

Reply:
Done